feat: Add detail to README + add createPlanets to local game (#8)
projectsophon/darkforest-local: README.md

# darkforest-local

> The Dark Forest client deploy process is set up for players to easily fork the client and connect to
> the mainnet game, but that design decision currently makes it difficult to run a playable version
> of the game locally. This repository provides a setup for running a local game with just a few
> steps.

Reviewer: ```suggestion
The Dark Forest client deploy process is set up for players to easily fork the client and connect to the mainnet game, but that design decision currently makes it difficult to run a playable version of the game locally. This repository provides a setup for running a local game with just a few steps.
```

> ## Builder's Guide
> 
> For a full tutorial on using `darkforest-local` with multiple **quests**, check out the [Community Round Builder's Guide](builders_guide.md).

Reviewer: ```suggestion
```

We'll add this section in the Builders Guide PR.

> ## How darkforest-local Works

Reviewer: ```suggestion
## How `darkforest-local` works
```

> This repo uses [submodules](.gitmodules) to pair the [Dark Forest Ethereum
> backend](https://github.com/darkforest-eth/eth) with the [Dark Forest TypeScript
> frontend](https://github.com/darkforest-eth/client) so you can launch a local game.

Reviewer: ```suggestion
This repo uses [submodules](.gitmodules) to pair the [Dark Forest Ethereum backend](https://github.com/darkforest-eth/eth) with the [Dark Forest TypeScript frontend](https://github.com/darkforest-eth/client) so you can launch a local game.
```

> It also uses Yarn, a package manager and that allows for multiple [workspaces](https://classic.yarnpkg.com/lang/en/docs/workspaces) to exist within a project.

Reviewer: ```suggestion
It also uses Yarn, a package manager (like `npm`) that allows for multiple [workspaces](https://classic.yarnpkg.com/lang/en/docs/workspaces) to exist within a project.
```

The workspaces allows each [submodule](.gitmodules) to have their own packages and configuration. 
Yarn places all of the packages for each submodule in the top level `node_modules/` folder.

## Requirements
* Install [Yarn](https://classic.yarnpkg.com/en/docs/install)

> * Install `node >= 14` [nvm](https://github.com/nvm-sh/nvm)

Reviewer: ```suggestion
* Install `node >= 14` (You can consider using [nvm](https://github.com/nvm-sh/nvm))
```

> ### Fastest Method for Running a Local Game

Reviewer: ```suggestion
### Quickstart for running a local game
```

> 1. Fork [darkforest-local](https://github.com/projectsophon/darkforest-local) to your Github repo.

Reviewer: ```suggestion
1. Fork [darkforest-local](https://github.com/projectsophon/darkforest-local) to your GitHub account
```

> 2. `git clone https://github.com/<your_name>/darkforest-local.git`

Reviewer: ```suggestion
2. `git clone --recurse-submodules https://github.com/<your_name>/darkforest-local.git`
```

> 3. `git submodule update --init --recursive --remote --merge`

Reviewer: ```suggestion
3. If you didn't clone with `--recurse-submodules` or already have a cloned version: `git submodule update --init --recursive --remote --merge`
```

4. `yarn`
5. `yarn start`

> ### Better Method for Running a Local Game

Reviewer: ```suggestion
### If you plan to make changes to `darkforest-local`
```

Author: I link to Better Method for Running a Local Game in the builder's guide so will need to update this

> 1. Fork [darkforest-local](https://github.com/projectsophon/darkforest-local) to your Github repo.
> 2. Fork [darkforest-eth/eth](https://github.com/darkforest-eth/eth) to your Github repo
> 3. Fork [darkforest-eth/client](https://github.com/darkforest-eth/client) to your Github repo.

Reviewer: ```suggestion
1. Fork [darkforest-local](https://github.com/projectsophon/darkforest-local) to your GitHub account
2. Fork [darkforest-eth/eth](https://github.com/darkforest-eth/eth) to your GitHub account
3. Fork [darkforest-eth/client](https://github.com/darkforest-eth/client) to your GitHub account
4. Fork [darkforest-eth/circuits](https://github.com/darkforest-eth/circuits) to your GitHub account
5. Fork [darkforest-eth/packages](https://github.com/darkforest-eth/packages) to your GitHub account
```

4. Clone your darkforest-local repo: `git clone https://github.com/<your_name>/darkforest-local.git`

> 5. Update the `.gitmodules` file to point to your new forks of `eth` and `client.  

Reviewer: ```suggestion
5. Update the `.gitmodules` file to point to your new forks of `eth`, `client`, `circuits`, and `packages` 
```

    ex: 

>     * `url = https://github.com/darkforest-eth/eth` =>  
>     `url = https://github.com/cha0sg0d/eth`
>     * `url = https://github.com/darkforest-eth/client` =>  
>     `url = https://github.com/cha0sg0d/client`

Reviewer: ```suggestion
    * `url = https://github.com/darkforest-eth/eth` => `url = https://github.com/cha0sg0d/eth`
    * `url = https://github.com/darkforest-eth/client` => `url = https://github.com/cha0sg0d/client`
    * `url = https://github.com/darkforest-eth/circuits` => `url = https://github.com/cha0sg0d/circuits`
    * `url = https://github.com/darkforest-eth/packages` => `url = https://github.com/cha0sg0d/packages`
```

6. Fetch the code from the submodules
    * `git submodule update --init --recursive`   
7. Add new branches for developing:
    - The `darkforest-local` monorepo detaches the submodules from their current HEADs. If you want to save your changes (for example, if you're testing an new contract in `eth`), you'll need to make a new branch in these submodules.
    1. `cd eth`
        1. `git checkout -b <new_name>`
    2. `cd client`
        1. `git checkout -b <new_name>`
7. Install packages and dependencies
    * `yarn`
8. Start a game
    * `yarn start`

        If `yarn start` returns the error `ERR_OSSL_EVP_UNSUPPORTED`:  
        * *This error occurs when using Node v17+*

>         * To fix, run `export NODE_OPTIONS=--openssl-legacy-provider` or switch to `14 <= node <= 16`

Reviewer: Do we need this? Odd version node aren't recommended for use, as they are canary builds that break all the time.

Author: good to know, let's remove

The Dark Forest client deploy process is set up for players to easily fork the client and connect to the mainnet game, but that design decision currently makes it difficult to run a playable version of the game locally. This repository provides a setup for running a local game with just a few steps.

## How `darkforest-local` works

This repo uses [submodules](.gitmodules) to pair the [Dark Forest Ethereum backend](https://github.com/darkforest-eth/eth) with the [Dark Forest TypeScript frontend](https://github.com/darkforest-eth/client) so you can launch a local game.

It also uses Yarn, a package manager (like `npm`) that allows for multiple [workspaces](https://classic.yarnpkg.com/lang/en/docs/workspaces) to exist within a project.

The workspaces allows each [submodule](.gitmodules) to have their own packages and configuration. 
Yarn places all of the packages for each submodule in the top level `node_modules/` folder.

## Requirements

> * Install [Yarn](https://classic.yarnpkg.com/en/docs/install)
> * Install `node >= 14` (You can consider using [nvm](https://github.com/nvm-sh/nvm))

Reviewer: ```suggestion

* Install `node >= 14` (You can consider using [nvm](https://github.com/nvm-sh/nvm))
* Install [Yarn](https://classic.yarnpkg.com/en/docs/install)
```

Let's reorder these.

5. Fork [darkforest-eth/packages](https://github.com/darkforest-eth/packages) to your GitHub account
4. Clone your darkforest-local repo: `git clone https://github.com/<your_name>/darkforest-local.git`
5. Update the `.gitmodules` file to point to your new forks of `eth`, `client`, `circuits`, and `packages` 
    ex: 
    * `url = https://github.com/darkforest-eth/eth` => `url = https://github.com/cha0sg0d/eth`
    * `url = https://github.com/darkforest-eth/client` => `url = https://github.com/cha0sg0d/client`
    * `url = https://github.com/darkforest-eth/circuits` => `url = https://github.com/cha0sg0d/circuits`
    * `url = https://github.com/darkforest-eth/packages` => `url = https://github.com/cha0sg0d/packages`
6. Fetch the code from the submodules
    * `git submodule update --init --recursive`   
7. Add new branches for developing:
    - The `darkforest-local` monorepo detaches the submodules from their current HEADs. If you want to save your changes (for example, if you're testing an new contract in `eth`), you'll need to make a new branch in these submodules.
    1. `cd eth`
        1. `git checkout -b <new_name>`

>     2. `cd client`
>         1. `git checkout -b <new_name>`

Reviewer: ```suggestion
    2. `cd client`
        1. `git checkout -b <new_name>`
    3. `cd circuits`
        1. `git checkout -b <new_name>`
    4. `cd packages`
        1. `git checkout -b <new_name>`
```
Probably should include the other submodules.

7. Install packages and dependencies
    * `yarn`
8. Start a game
    * `yarn start`

>         If `yarn start` returns the error `ERR_OSSL_EVP_UNSUPPORTED`:  
>         * *This error occurs when using Node v17+*
>         * To fix, run `export NODE_OPTIONS=--openssl-legacy-provider` or switch to `14 <= node <= 16`

Reviewer: ```suggestion
```

`client/` and `eth/` submodules. You can find the set of environment variables you will need to
populate those `.env` files with in their adjacent `.env.example` files.

> > Danger! The `.env` files ARE in the respective `.gitignore`s of the aforementioned submodules,

Reviewer: ```suggestion
> **Danger!** The `.env` files ARE in the respective `.gitignore`s of the aforementioned submodules,
```

![yarn workspace eth hardhat:prod deploy](img/hardhat_prod_deploy.png)

To find out more information about the `contracts` submodule, you can look here:
[https://github.com/darkforest-eth/eth](https://github.com/darkforest-eth/eth).

### Step 2: Deploy client website

To deploy the website interface, you may either self-host, or use the same infra that we use -
[Netlify](https://www.netlify.com/). This is a simple option, and free for up to some amount of
gigabytes of bandwidth per month, which has often been enough for us.

To use Netlify:

> - Make a new [Netlify account](https://app.netlify.com/signup) using your Github profile.

Reviewer: ```suggestion
- Make a new [Netlify account](https://app.netlify.com/signup) using your GitHub account
```

> - [Import](https://app.netlify.com/start) your Github client repo as a new Netlify site.

Reviewer: ```suggestion
- [Import](https://app.netlify.com/start) your `client` repository as a new Netlify site.
```

- Install the Netlify CLI
    - `npm install netlify-cli -g`
- Login to your account
    - `netlify login`
- Connect to your app
    - `netlify link`
    

>     ![link](img/link.png)

Reviewer: ```suggestion
    ![Netlify link](img/link.png)
```

    
- Choose *Enter a site ID*
    

>     ![Screen Shot 2022-01-15 at 12.35.15 PM.png](img/site_id.png)

Reviewer: ```suggestion
    ![Site ID](img/site_id.png)
```

    
- Input the id that you find by clicking on your new site on Netlify. In this example, the id is `jolly-hermann-3947ca`
- Now that your site is linked, you can deploy to Netlify from the CLI.
    - `yarn workspace client deploy:prod`
    - If successful, you will see something like this.
        
        ![deploy_success](img/netlify_deploy.png)
        

>     - *Note: To do this, you will have to turn on automatic deploys for your site. This means that if you push an update to your main Git branch, Netlify will attempt to redeploy the site. This will fail, because the client relies on the most recent versions of the packages you have created, but Netlify will try to install the older versions from npm instead of locally.*

Reviewer: Why don't you turn off the deploys? Or instead recommend the users to deploy a subdirectory of darkforest-local?

>         ![Screen Shot 2022-01-15 at 12.38.57 PM.png](img/prod_deploy.png)

Reviewer: ```suggestion
        ![deploy_failures](img/prod_deploy.png)
```

![enter game](img/enter_game.png)

### Step 4: Play your game

- Now go the URL given by Netlify → something like [https://jolly-hermann-3947ca.netlify.app/](https://jolly-hermann-3947ca.netlify.app/) and test out your game! Login with one of the private keys for the addresses that were whitelisted. Make some moves and test that everything works as expected.

## Update to latest Dark Forest code

> If theres been new Dark Forest updates released and we haven't yet updated this repo yet, it is
> possible you can get away with updating yourself by running `git submodule update --remote --merge`
> and remember to run `yarn` again.

Reviewer: ```suggestion
If theres been new Dark Forest updates released and we haven't yet updated this repo yet, it is possible you can get away with updating yourself by running `git submodule update --remote --merge` and remember to run `yarn` again.
```
